### PR TITLE
added CarService, CarDao, Car. Reformed packages.

### PR DESCRIPTION
**Pull request by Daniel1004K**

### CODE REVIEW

File: src/main/java/mate/jdbc/Main.java [hunk 1]
```diff
@@ -1,7 +1,42 @@
...
+
     public static void main(String[] args) {
-        // test your code here
+        Driver driver = new Driver();
+        driver.setName("ABCD");
+        driver.setLicenseNumber("A-B-C-D");
+        driverService.create(driver);
+        Car car1 = carDao.get(1L);
+        carService.addDriverToCar(driver, car1);
+        carDao.getAll().forEach(System.out::println);
```
Comment (deniskichan) [inline src/main/java/mate/jdbc/Main.java:29]: we shouldn't use DAO level in our main method

File: src/main/java/mate/jdbc/dao/impl/CarDaoImpl.java [hunk 2]
```diff
@@ -0,0 +1,249 @@
...
+                    e);
+        }
+        return driverList;
+    }
+
+    private Driver parseDriverFromResultSet(ResultSet driversResultSet) throws SQLException {
+        Driver driver = new Driver();
+        driver.setName(driversResultSet.getString("name"));
+        driver.setLicenseNumber(driversResultSet.getString("license_number"));
+        driver.setId(driversResultSet.getLong("id"));
```
Comment (deniskichan) [inline src/main/java/mate/jdbc/dao/impl/CarDaoImpl.java:230]:
```suggestion
        driver.setId(driversResultSet.getObject("id", Long.class));
```

File: src/main/java/mate/jdbc/dao/impl/CarDaoImpl.java [hunk 3]
```diff
@@ -0,0 +1,249 @@
...
+        return driver;
+    }
+
+    private Car parseCarFromResultSet(ResultSet carsResultSet) throws SQLException {
+        Car car = new Car();
+        car.setId(carsResultSet.getLong("id"));
```
Comment (deniskichan) [inline src/main/java/mate/jdbc/dao/impl/CarDaoImpl.java:236]: same, better use getObject

File: src/main/java/mate/jdbc/dao/impl/CarDaoImpl.java [hunk 4]
```diff
@@ -0,0 +1,249 @@
...
+        car.setModel(carsResultSet.getString("model"));
+        return car;
+    }
+
+    private Manufacturer parseManufacturerFromResultSet(ResultSet manufacturerResultSet)
+            throws SQLException {
+        Manufacturer manufacturer = new Manufacturer();
+        manufacturer.setId(manufacturerResultSet.getLong("id"));
```
Comment (deniskichan) [inline src/main/java/mate/jdbc/dao/impl/CarDaoImpl.java:244]: same, better use getObject

**Review by deniskichan — APPROVED** [review]

File: src/main/java/mate/jdbc/dao/CarDao.java [hunk 5]
```diff
@@ -0,0 +1,18 @@
+package mate.jdbc.dao;
+
+import java.util.List;
+import mate.jdbc.model.Car;
+
+public interface CarDao {
+    Car create(Car car);
+
+    Car get(Long id);
```
Comment (Sofasmile) [inline src/main/java/mate/jdbc/dao/CarDao.java:9]:
```suggestion
    Optional<Car> get(Long id);
```

File: src/main/java/mate/jdbc/dao/impl/CarDaoImpl.java [hunk 6]
```diff
@@ -0,0 +1,249 @@
...
+            ResultSet carResultSet = getCarStatement.executeQuery();
+            if (carResultSet.next()) {
+                car = parseCarFromResultSet(carResultSet);
+            }
+        } catch (SQLException e) {
+            throw new DataProcessingException("Can't get car by id = " + id + " from DB taxi.cars",
+                    e);
+        }
+        if (car != null) {
+            car.setManufacturer(getManufacturerForCar(id));
```
Comment (Sofasmile) [inline src/main/java/mate/jdbc/dao/impl/CarDaoImpl.java:59]: can we get car and manufacturer using one SQL statement?

File: src/main/java/mate/jdbc/dao/impl/CarDaoImpl.java [hunk 7]
```diff
@@ -0,0 +1,249 @@
...
+        try (Connection connection = ConnectionUtil.getConnection();
+                PreparedStatement getManufacturerStatement = connection
+                        .prepareStatement(getManufacturerForCar)) {
+            getManufacturerStatement.setLong(1, id);
+            ResultSet driversResultSet = getManufacturerStatement.executeQuery();
+            while (driversResultSet.next()) {
+                driverList.add(parseDriverFromResultSet(driversResultSet));
+            }
+        } catch (SQLException e) {
+            throw new DataProcessingException("Can't get manufacturer from DB taxi.manufacturers",
```
Comment (Sofasmile) [inline src/main/java/mate/jdbc/dao/impl/CarDaoImpl.java:220]:
```suggestion
            throw new DataProcessingException("Can't get manufacturer from DB taxi.manufacturers by car_id " + id,
```

File: src/main/java/mate/jdbc/dao/impl/CarDaoImpl.java [hunk 8]
```diff
@@ -0,0 +1,249 @@
...
+            while (carsResultSet.next()) {
+                carList.add(parseCarFromResultSet(carsResultSet));
+            }
+        } catch (SQLException e) {
+            throw new DataProcessingException("Can't get cars from DB taxi.cars", e);
+        }
+        if (!carList.isEmpty()) {
+            for (Car car : carList) {
+                car.setDrivers(getDriversForCar(car.getId()));
+                car.setManufacturer(getManufacturerForCar(car.getId()));
```
Comment (Sofasmile) [inline src/main/java/mate/jdbc/dao/impl/CarDaoImpl.java:81]: same here with Manufacturer, let's get it with car using one SQL query

File: src/main/java/mate/jdbc/dao/impl/CarDaoImpl.java [hunk 9]
```diff
@@ -0,0 +1,249 @@
...
+        try (Connection connection = ConnectionUtil.getConnection();
+                Statement getAllCarsStatement = connection.createStatement()) {
+            ResultSet carsResultSet = getAllCarsStatement.executeQuery(getAllCarsRequest);
+            while (carsResultSet.next()) {
+                carList.add(parseCarFromResultSet(carsResultSet));
+            }
+        } catch (SQLException e) {
+            throw new DataProcessingException("Can't get cars from DB taxi.cars", e);
+        }
+        if (!carList.isEmpty()) {
```
Comment (Sofasmile) [inline src/main/java/mate/jdbc/dao/impl/CarDaoImpl.java:78]: do we need this checking?

File: src/main/java/mate/jdbc/dao/impl/CarDaoImpl.java [hunk 10]
```diff
@@ -0,0 +1,249 @@
...
+            }
+        }
+        return carList;
+    }
+
+    @Override
+    public Car update(Car car) {
+        String updateCarRequest = "UPDATE cars SET model = ?, manufacturer_id = ? "
+                + "WHERE id = ? AND is_deleted = FALSE";
+        boolean changeDrivers;
```
Comment (Sofasmile) [inline src/main/java/mate/jdbc/dao/impl/CarDaoImpl.java:91]:
```suggestion
        boolean shouldChangeDrivers;
```

File: src/main/java/mate/jdbc/dao/impl/CarDaoImpl.java [hunk 11]
```diff
@@ -0,0 +1,249 @@
...
+
+    private boolean deleteDriversCar(Car car) {
+        String deleteAllDriversByCarRequest = "DELETE FROM cars_drivers WHERE car_id = ?";
+        try (Connection connection = ConnectionUtil.getConnection();
+                PreparedStatement deleteDriversStatement = connection
+                        .prepareStatement(deleteAllDriversByCarRequest)) {
+            deleteDriversStatement.setLong(1, car.getId());
+            return deleteDriversStatement.executeUpdate() != 0;
+        } catch (SQLException e) {
+            throw new DataProcessingException("Can't deleted drivers by car in DB cars_drivers", e);
```
Comment (Sofasmile) [inline src/main/java/mate/jdbc/dao/impl/CarDaoImpl.java:163]:
```suggestion
            throw new DataProcessingException("Can't deleted drivers by car " + car + " in DB cars_drivers", e);
```

File: src/main/java/mate/jdbc/dao/impl/CarDaoImpl.java [hunk 12]
```diff
@@ -0,0 +1,249 @@
...
+            throw new DataProcessingException("Can't delete car by id = " + id, e);
+        }
+    }
+
+    @Override
+    public List<Car> getAllByDriver(Long driverId) {
+        List<Car> carList = new ArrayList<>();
+        String getAllCarByDriverRequest =
+                "SELECT c.id,c.model, c.manufacturer_id "
+                        + "FROM taxi.cars_drivers cd "
```
Comment (Sofasmile) [inline src/main/java/mate/jdbc/dao/impl/CarDaoImpl.java:128]:
```suggestion
                        + "FROM cars_drivers cd "
```

File: src/main/java/mate/jdbc/dao/impl/CarDaoImpl.java [hunk 13]
```diff
@@ -0,0 +1,249 @@
...
+        } catch (SQLException e) {
+            throw new DataProcessingException("Can't delete car by id = " + id, e);
+        }
+    }
+
+    @Override
+    public List<Car> getAllByDriver(Long driverId) {
+        List<Car> carList = new ArrayList<>();
+        String getAllCarByDriverRequest =
+                "SELECT c.id,c.model, c.manufacturer_id "
```
Comment (Sofasmile) [inline src/main/java/mate/jdbc/dao/impl/CarDaoImpl.java:127]: same about the manufacturer. Let's get it here

File: src/main/java/mate/jdbc/service/impl/CarServiceImpl.java [hunk 14]
```diff
@@ -0,0 +1,52 @@
...
+    public Car getCarById(Long id) {
+        return carDao.get(id);
+    }
+
+    @Override
+    public void addDriverToCar(Driver driver, Car car) {
+        List<Driver> drivers = car.getDrivers();
+        drivers.add(driver);
+        car.setDrivers(drivers);
+        carDao.update(car);
```
Comment (Sofasmile) [inline src/main/java/mate/jdbc/service/impl/CarServiceImpl.java:26]:
```suggestion
        car.getDrivers().add(driver);
        carDao.update(car);
```

File: src/main/java/mate/jdbc/service/impl/CarServiceImpl.java [hunk 15]
```diff
@@ -0,0 +1,52 @@
...
+    }
+
+    @Override
+    public void removeDriverFromCar(Driver driver, Car car) {
+        List<Driver> drivers = car.getDrivers();
+        if (drivers.remove(driver)) {
+            car.setDrivers(drivers);
+            carDao.update(car);
```
Comment (Sofasmile) [inline src/main/java/mate/jdbc/service/impl/CarServiceImpl.java:34]:
```suggestion
            carDao.update(car);
```

**Review by Sofasmile — CHANGES_REQUESTED** [review]
great job)

File: src/main/java/mate/jdbc/dao/impl/CarDaoImpl.java [hunk 16]
```diff
@@ -124,10 +129,12 @@ public boolean delete(Long id) {
     public List<Car> getAllByDriver(Long driverId) {
         List<Car> carList = new ArrayList<>();
         String getAllCarByDriverRequest =
-                "SELECT c.id,c.model, c.manufacturer_id "
-                        + "FROM taxi.cars_drivers cd "
-                        + "JOIN cars c on cd.car_id = c.id "
-                        + "WHERE cd.driver_id = ? AND c.is_deleted = FALSE;";
+                "SELECT * "
+                        + "FROM cars_drivers AS cd, manufacturers AS m, cars AS c "
```
Comment (Sofasmile) [inline src/main/java/mate/jdbc/dao/impl/CarDaoImpl.java:133]: чому ми тут не використовуємо join?

File: src/main/java/mate/jdbc/dao/impl/CarDaoImpl.java [hunk 17]
```diff
@@ -227,23 +207,19 @@ private Driver parseDriverFromResultSet(ResultSet driversResultSet) throws SQLEx
         Driver driver = new Driver();
         driver.setName(driversResultSet.getString("name"));
         driver.setLicenseNumber(driversResultSet.getString("license_number"));
-        driver.setId(driversResultSet.getObject("id",Long.class));
+        driver.setId(driversResultSet.getObject("id", Long.class));
         return driver;
     }
 
     private Car parseCarFromResultSet(ResultSet carsResultSet) throws SQLException {
```
Comment (Sofasmile) [inline src/main/java/mate/jdbc/dao/impl/CarDaoImpl.java:214]:
```suggestion
    private Car parseCarWithManufacturerFromResultSet(ResultSet carsResultSet) throws SQLException {
```

File: src/main/java/mate/jdbc/service/impl/CarServiceImpl.java [hunk 18]
```diff
@@ -0,0 +1,50 @@
...
+    public void addDriverToCar(Driver driver, Car car) {
+        car.getDrivers().add(driver);
+        carDao.update(car);
+    }
+
+    @Override
+    public void removeDriverFromCar(Driver driver, Car car) {
+        List<Driver> drivers = car.getDrivers();
+        if (drivers.remove(driver)) {
+            car.setDrivers(drivers);
```
Comment (Sofasmile) [inline src/main/java/mate/jdbc/service/impl/CarServiceImpl.java:31]: how about this comment? do we need to set drivers to the car?

**Review by Sofasmile — CHANGES_REQUESTED** [review]

File: src/main/java/mate/jdbc/dao/impl/CarDaoImpl.java [hunk 19]
```diff
@@ -129,19 +129,19 @@ public boolean delete(Long id) {
...
-                        + "WHERE c.id = ? "
-                        + "AND c.is_deleted = FALSE "
-                        + "AND m.id = c.id "
-                        + "AND cd.car_id = c.id ;";
+                "SELECT tab.car_id, tab.model, manufacturer_id, m.name, m.country "
+                        + "FROM manufacturers AS m "
+                        + "INNER JOIN ("
+                        + "SELECT * FROM cars "
+                        + "INNER JOIN cars_drivers ON cars.id = cars_drivers.car_id"
+                        + ") as tab ON m.id = tab.car_id;";
```
Comment (Daniel1004K) [inline src/main/java/mate/jdbc/dao/impl/CarDaoImpl.java:137]: double join =)

**Review by Sofasmile — APPROVED** [review]

**Pull request closed without merging**